drivers/adc_ng/include/adc_ng_internal.h

```diff
@@ -0,0 +1,289 @@
+ */
+#define ADC_NG_REF_MCU_VCC              (0x8000U)
+
+/**
+ * @brief   Flag a reference voltage as calibrated
+ *
+ * E.g. the 1.1V bandgap voltage reference available on many ATmega has
+ * production variations of +-100mV, but is extremely stable; so it will
+ * consistently create the same voltage for a wide range of supply voltage and
+ * operating temperatures. So if the actual value of the reference is e.g.
+ * written to EEPROM and restored on boot, the reference value should be marked
+ * as calibrated.
+ */
+#define ADC_NG_REF_CALIBRATED           (0x4000U)
+/**
+ * @brief   Access the voltage value without flags
+ */
+#define ADC_NG_REF_MASK                 (0x3fffU)
+
+/**
+ * @brief   Description of an thermistor to use for temperature measurements
+ */
+typedef struct {
+    /**
+     * @brief   Contains the temperature coefficient of the NTC, or zero
+     *
+     * The coefficient is is given in 1/1024 mV per 0.1 °C. The resulting
+     * temperature in 0.1 °C is calculated from the measured voltage using:
+     *
+     *    T[d°C] = (coefficient * (mV - offset)) / 1024
+     */
+    uint16_t coefficient;
+    /**
+     * @brief   The offset in mV to use for obtaining the temperature
+     *
+     * See @ref adc_ng_ntc_t::coefficient
+     */
+    uint16_t offset;
+} adc_ng_ntc_t;
+
+/**
+ * @brief   Internal driver interface
+ */
+typedef struct {
+    /**
+     * @brief   Initialize the given ADC channel and prepare it for sampling
+     * @param           handle  Handle of the ADC
+     * @param[in]       chan    The ADC channel to initialize
+     * @param[in]       res     The resolution to select
+     * @param[in        ref     Index of the reference to use (@ref adc_ng_driver_t::refs)
+     *
+     * @retval  0               Success
+     * @retval  -ENXIO          Invalid channel given in @p channel
+     * @retval  -EALREADY       The ADC is already powered and configured
+     * @retval  <0              Other error (see device driver doc)
+     *
+     * @post    When 0 is returned, the channel @p channel is ready take samples
+     * @post    If @p res contains an unsupported resolution, an assert blows up
+     * @post    If `-EALREADY` is returned, the ADC is state remains unchanged
+     * @post    On other error codes the ADC is powered down
+     *
+     * @note    A call to @ref adc_ng_driver_t::off is needed to disable the channel
+     *          again and preserve power
+     */
+    int (*init)(void *handle, uint8_t chan, uint8_t res, uint8_t ref);
+    /**
+     * @brief   Disable the given ADC channel again and bring the ADC into a low
+     *          power state, unless other ADC channels are still onchannel
+     *
+     * @param           handle  Handle of the ADC
+     */
+    void (*off)(void *handle);
+    /**
+     * @brief   Runs a single conversion and returns the sample
+     *
+     * @param           handle  Handle of the ADC
+     * @param[out]      dest    The sample will be stored here
```

Comment:
How to deal with an ADC that uses a FSR of +-6.144V and provides samples as two's complement? That is the reference voltage as well as the samples can be negative.

Reply:
IMHO, a possible solution would be to specify a reference voltage of 12,288 mV and to convert the to `unsigned` so that 0 means -6.144V and 65,535 means +6.144V for 16 bit resolution.

Reply:
Why not just use a signed type for samples then?

Reply:
Because the API specifies an `unsigned` sample type.

Reply:
But… this PR is about discussing a new API.

Reply:
I guess it would certainly not hurt to have the sample as `int32_t` instead of `uint32_t`. That would of course limit use to "only" support up to 31 bit resolution; but an internal implementation details does so anyway. (Of course this detail could be easily changed without touching the user facing API.) In any case: For practical reasons there has to be an upper limit on the resolution and 31 bit is to me just as good as 32 bit, and having an easy way to also encode negative value seems to be of more value than one bit of resolution more.

E.g. the ATmega ADC on the bigger MCUs can use a differential mode with two inputs that potentially returns negative values; so negative values are not that uncommon.

I also think that an `int16_t` for the reference voltage with a range of -32.768V to 32.767V would be an issue. I don't expect reference voltages of over 32V to be supported by devices running RIOT anyway.

Reply:
But there is one thing that should be agreed on. E.g. lets say we have a resolution of 10 bits and a reference voltage of -1.023V. The input voltage is -0.512V. Should the sample now have the value 512 or -512?

Reply:
> I guess it would certainly not hurt to have the sample as int32_t instead of uint32_t. That would of course limit use to "only" support up to 31 bit resolution; but an internal implementation details does so anyway. 

I think it will become a bit more complicated for 32-bit ADCs. For example, the samples from an [ADS1262](https://www.ti.com/product/ADS1262) are given as two's complement. That is, an `int32_t` type would be perfect. But the samples from an [AD177-2](https://www.analog.com/media/en/technical-documentation/data-sheets/AD7177-2.pdf) are either given as bipolar offset values or as unsigned values. In latter case, only positive inputs are allowed but in full 32-bit range. `bipolar offset` values could be converted to `int32_t` values.

> The input voltage is -0.512V. Should the sample now have the value 512 or -512?

Good question. According to `adc_ng_convert` it should be 512, right?

Reply:
To me 512 would also make most sense. But as long as we specify this in the API, so that all ADC drivers used over the common API behave consistently, I actually wouldn't mind either.

Reply:
When specifying a negative reference voltage, e.g. `-2.0V` and supported is `1.8V` and `-2.56V` the voltage `-2.56V` should be chosen in order to cover the requested voltage range, right?

I will have to update the documentation and implementation a bit.

Reply:
> When specifying a negative reference voltage, e.g. -2.0V and supported is 1.8V and -2.56V the voltage -2.56V should be chosen in order to cover the requested voltage range, right?

I mean you are always measuring against the reference voltage, not against ground.
So -2.56V would be zero (or the maximum value) and 1.8V against ground would be 4.36V against the reference (or 0/max in raw values depending on which way round you are measuring)

Reply:
> I mean you are always measuring against the reference voltage, not against ground.
So -2.56V would be zero.

That is `uint32_t` would be fine in any case?

Reply:
Not if we want to add support for differential channels with to pins as inputs. The a negative value is used to indicate that the positive input has a lower voltage than the negative input. The ATmega ADC on the bigger MCUs supports this.

Reply:
From what I understand yes.

Reply:
> differential channels

External ADC devices supports this usually too.

Reply:
See above ;-)

Reply:
I changed both results and reference voltage types to be signed. The documentation states that single ended inputs should have no sign, so that no misunderstanding and inconsistencies between different implementation should sneak in. (But I should check the reference voltage selection code again if this works correct for negative reference voltages.)